Prevent sending more emails than allowed [MAILPOET-4122]

websupporter opened the pull request:
Fixes [MAILPOET-4122]

Before this PR, MailPoet was able to send more emails than the settings allowed for.

# A conceptual problem
Before this PR, we calculated whether the sending limit was reached in a data model, which consisted out of the number of emails sent since a fixed timestamp. So, once we started a new "series", we would set the timestamp and then just count up until the limit was reached. Once the limit was reached, we would check if the timestamp we had was outdated, in which case we would reset the number of emails, set a new timestamp and continue sending.

This model had some flaws:
1. Example: Limit is 100 emails, allowed timespan is 1 minute. Lets say our last "series" consisted of 50 emails. This left 50 unused "spots". Before we would check whether the timestamp was outdated (hence, we were in a new timespan), we would fill those spots. Then we would create a new timespan and would allow for 100 more emails, although we already did send 50.
2. We had no idea about the distribution of the emails, which we send. Theoretically this could lead to a timespan, where one email was send at the beginning and 99 in the last second. Although our timespan was filled and we would restart, this could lead to sending more than 100 emails in a minute.

This PR suggests a different approach: A "rolling timeframe". Each second we send an email would create a new entry in an array with the number of emails we did send in this second. When we calculate how many emails where send in the last 60 seconds, we `array_sum()` all entries in this timeframe. Once entries are outdated and fall out of the timeframe, we delete them. With the current UI, we can set the sending timeframe up to 30 minutes, which leaves us with max. 1800 entries in this array. A "rolling timeframe" keeps the information about distribution and therefore tackles the two flaws of the previous model. ([23d09b5](https://github.com/mailpoet/mailpoet/pull/4001/commits/23d09b59f0ce3a5637b47e6c2dc0f1c4d7a8151c))

~A problem of keeping track~
~A second problem existed, as we did not count all emails, which we did send. MailPoet offers the possibility to send all transactional emails. If you set a limit on how many emails you want to send in a given timeframe, those emails need to be accounted for. They weren't. This PR will~
~1. Enforce the sending execution requirements in the `Mailer::send()` method. So far, this requirements have been only checked in the SendingQueue, which does not go through the `Mailer::send()` method. Transactional emails and other emails are send via `Mailer::send()`. Besides the sending limit the execution requirements also count for sending paused or blocked by a retry waiting period.~
~2. Increment the count of the emails send, when they are send using `Mailer::send()`.~

~What happens when an transactional email can not be send?~
~At the moment, MailPoet has no queue for transactional emails. If an error occurs, MailPoet falls back to the default PHPMailer and attempts to send the email. This behavior remains.~
~(6a23fb4)~

~A non-static MailerLog~
~The MailerLog class so far was a static class. This PR rewrote the class and got rid of its static attributes (bf5d41fd1e121d586516bc54a6c334f4f42afe76)~

# Introduce a phpstan-type MailerLogData
The data with which the MailerLog works is a simple array. The array stores the data whether the sending is paused, what errors occurred and so on. As it is just an array and not an object/entity, this can easily lead to errors, we would like to prevent (e.g. an array key is expected to be present, but it isn't and therefore notices are thrown). This PR typecasts most methods and introduces an phpstan-type, so that PHPStan can easily detect missing or misspelled array keys.

(83846fc)

[MAILPOET-4122]: https://mailpoet.atlassian.net/browse/MAILPOET-4122?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

rodrigoprimo commented:
Thanks for the detailed description of the problem that is fixed by this PR, @websupporter.

I have a few questions before I start the review:

- Is there a way to reproduce the problem that this PR fixes so that I can easily confirm that the fix works as expected? I understand that sometimes there isn't, and it might be the case here.

> The MailerLog class so far was a static class. This PR rewrote the class and got rid of its static attributes

Could you please elaborate on why this change is needed? The fact that the class was static is part of the reason why this problem was happening? If not, could this refactor be done in a separate PR so that it is easier to see here only the changes that are required for this fix? Just trying to make sure we have here only the things that are needed for this fix as this is a big PR (which sometimes is necessary and that is ok).

Similar to the above, maybe commit 9d4d349ea9e8600f47f00898beb015d56eb71013 can be part of a separate PR or statements became outdated because of changes in this PR?

websupporter commented:
Thank you @rodrigoprimo.

Tbh [9d4d349](https://github.com/mailpoet/mailpoet/pull/4001/commits/9d4d349ea9e8600f47f00898beb015d56eb71013) was a local rebase of master going south.

Regarding `static`: It is not immediately necessary. My reasoning was to improve testability of the class itself and further improve our ability to do unit testing, as `MailLog` now needs to be injected and therefore can be mocked. It can be separated into another PR.

Regarding testing:
1. Checkout master
2. Create a list with 150 subscribers
3. Set your sending limit to 100/5 minutes
4. Send an email to all subscribers
5. (The internal state auf the MailLog data should now be `sent => 50, started => some-timestamp`
6. Wait a bit, so you are out of the 5 minutes timeframe
7. Send again an email to all 150 subscribers
8. You should be able to send all 150 emails in under five minutes (due to the bug)
9. Checkout fix/send-more-emails-than-allowed-mailpoet-4122
10. Repeat
11. You should not be able to send all 150 emails in under five minutes (due to the fix of the bug :crossed_fingers:)

Regarding also counting transactional emails:
[0e939a0](https://github.com/mailpoet/mailpoet/pull/4001/commits/0e939a07d7a61bacb2cc7861b60af44021370b4b) contains an integration test which shows that the sending fails, when you reach the limit, [327ceaa](https://github.com/mailpoet/mailpoet/pull/4001/commits/327ceaa25fd88eebff56deb859bf01a8c1984135) adds a test to demonstrate that the count how many emails have been send gets incremented, when a mail is send with the `Mailer`.
 
1. Shall I separate those PRs
2. Should the phpstan-type be a separate PR as well?

rodrigoprimo commented:
Thanks for the additional details and for the testing instructions, David.

> Tbh [9d4d349](https://github.com/mailpoet/mailpoet/pull/4001/commits/9d4d349ea9e8600f47f00898beb015d56eb71013) was a local rebase of master going south.

:+1: 

> Regarding static: It is not immediately necessary. My reasoning was to improve testability of the class itself and further improve our ability to do unit testing, as MailLog now needs to be injected and therefore can be mocked. It can be separated into another PR.
> (...)
> Shall I separate those PRs

Generally speaking, I agree with your motivation to change the class, but if it is not required to address the problem that is fixed here, I would suggest doing it in a separate PR. This should reduce the chances of us missing a potential problem either in the fix to the limit of allowed emails or the refactor. What do you think?

> Should the phpstan-type be a separate PR as well?

I'm ok with keeping the phpstan-type in this PR. It is a much smaller change, and my first impression without digging too much into the code is that it is directly related to the changes proposed here as it basically defines a type that is created in this PR and is at the core of the fix.

websupporter commented:
Contains now the fix for the issue and the phpstan doctype. Will prepare a 2nd PR for the question of a non-static MailerLog

rodrigoprimo commented on mailpoet/lib/Mailer/MailerLog.php:
```diff
@@ -129,54 +196,128 @@ public static function setError($mailerLog, $operation, $errorMessage, $errorCod
+    $sinceDate = self::sentEntriesDate(time() - $frequencyInterval);
+    $mailerLog = self::getMailerLog($mailerLog);
+
+    $mailerLog['sent'] = array_filter(
```
Same as above

rodrigoprimo reviewed (changes requested):
Good work, @websupporter. I left a few comments with some things for us to discuss and also I'm having an issue described below that, if confirmed, we need to fix before moving this forward.

As a sidenote, I was not able to reproduce the original problem on the master branch. Probably something that I'm doing wrong as both you and Jan Lysy were able to reproduce it.

withinboredom commented on mailpoet/lib/Mailer/MailerLog.php:
```diff
@@ -129,54 +196,128 @@ public static function setError($mailerLog, $operation, $errorMessage, $errorCod
+    $sinceDate = date('Y-m-d H:i:s', time() - $sinceSeconds);
+    $mailerLog = self::getMailerLog($mailerLog);
+
+    return (int)array_sum(
```
Hi :wave:,

Just visiting this repo and apparently subscribed to all PRs... But have you considered a "leaky bucket" algorithm which only requires you to keep track of a single int and the last time you put something in the bucket (sent an email)?

websupporter replied:
Hi @withinboredom,
thanks for chiming in. I am not sure yet about your suggestion. My first impression is, if we take the last time we sent an email, we do not know, when this timespan started. So if we say we want to send 100/5 minutes, once we reached 100, we would need to wait 5 minutes before we can start again, so we could only send 100/10 minutes. But probably I miss the point. Can you elaborate more?

JanJakes replied:
Hi, @withinboredom! Did you by any chance mean to say "[token bucket](https://en.wikipedia.org/wiki/Token_bucket)" instead of "leaky bucket"?

Leaky bucket leaks at a constant rate and needs a buffer (FIFO queue) for that. For us, this would mean always sending at a constant rate, which is not what we do or want to do.

The token bucket algorithm adds tokens at a constant rate and can use them variably. It will spread out the sending over the period uniformly, which is again not something we need or want to do.

I guess we need a throttling based on a "sliding window" algorithm, which is what this PR implements. For the sliding window, the best is to have an exact timestamp of every sent email, which we currently don't have (some transactional emails) so we need to store the counts. We can take it as a temporary solution later to be replaced by full sent email log.

websupporter commented on mailpoet/lib/Mailer/MailerLog.php:
```diff
@@ -4,25 +4,53 @@
+     * The old "sent" entry was just the number of emails.
+     * We need to update this entry to the new data structure.
+     */
+    $mailerLog['sent'] = is_numeric($mailerLog['sent']) ? [self::sentEntriesDate(time() - 1) => $mailerLog['sent']] : (array)$mailerLog['sent'];
```
Hello @rodrigoprimo,
thanks! I forgot to check that and it slipped through. I address it in 98f50e6b7

rodrigoprimo replied:
Thanks, @websupporter. I can confirm that the page is now working. I wonder if there are other places that rely on the format of `$mailLog['sent']` that could be broken as well? I don't recall anything off the top of my head, but maybe worth checking in case you haven't already?

websupporter commented:
Hi @rodrigoprimo,

thanks for your review. I digged a bit deeper into the performance of this PR. Obviously, this PR would have some performance impact. 

I did run the following test script against `master` and this PR:
```
require_once dirname(__DIR__,2) . '/wp-load.php';
$target = 50;
$counter = 0;
while ($counter < $target && ! MailerLog::isSendingLimitReached()) {
  echo $counter . PHP_EOL;
  MailerLog::incrementSentCount();
  $counter++;
}
```

I ensured the sending limit would never be reached. For the PR, I made sure that there would always be more than 1800 entries in the `sent` array. I tested it with several iterations (like 10.000 e.g.), but concluded the next to a real world example but still measurable would be to see the outcome for 50 mails.

Here is an overview over the results:
|                              	| normal visit of home page to compare 	| master    	| pr        	|
|------------------------------	|--------------------------------------	|-----------	|-----------	|
| total time                   	| 81698667                             	| 116717457 	| 224691133 	|
| total memory                 	| 59693264                             	| 118046664 	| 132322216 	|
| incrementSentCount time      	| ---                                  	| 43.87     	| 50.05     	|
| incrementSentCount memory    	| ---                                  	| 0.09      	| 2.34      	|
| isSendingLimitReached time   	| ---                                  	| 1.27      	| 19.05     	|
| isSendingLimitReached memory 	| ---                                  	| 0.00      	| 0.00      	|

Our batch sizes do not exceed 20 emails, so I tested, if the memory we use stays below 128 MB, as this is often a value for the memory limit. The test script itself with 1800+ entries reached 118mb, The master reached 117mb, so under normal circumstances, we do not really come close to the limits often imposed on memory usage.

So, to me: I do not expect the maximum expected size to cause problems under normal cicumstances. Still, we could reduce the maximum amount of entries in the `sent` array. 300 mails/30 minutes is effectifly 10 mails/minute. So we could transform those values and thus the max entries in the `sent` array. We would need to figure out how to handle the rest (e.g. 379/30=12.633) The downside risk here is, we rely on an assumption how often the cron is executed. I checked at least the server side cron. The suggested value to execute this cron remains one minute, regardless of the value I set as maximum. But as an administrator, I might use the setting that visitors execute the cron and due to low visit numbers, I decided to go with the 30 minutes option.

> I think we will be running an update query to update wp_mailpoet_settings for every email?

Yes, I didn't change that part of the logic and so this would extend now to all transactional emails as well. I checked mixpanel to see, if we can gain any insights into how many transactional emails are send, but we do not (probably can't right now) track that. For me, right now, it does not feel to worrysome, but this might be a lack of experience on the plugin. What we can immediately do: As MSS does not use the counter, we could stop the SQL write, when the sending method is MSS.

rodrigoprimo commented:
Thanks for looking deeper into the performance side of things.

> For the PR, I made sure that there would always be more than 1800 entries in the sent array.

My main concern was the size of the array. I didn't realize that the maximum number of entries was 1800 which doesn't seem a problem at all.

> The master reached 117mb, so under normal circumstances, we do not really come close to the limits often imposed on memory usage.

:+1: 

> Yes, I didn't change that part of the logic and so this would extend now to all transactional emails as well. I checked mixpanel to see, if we can gain any insights into how many transactional emails are send, but we do not (probably can't right now) track that. For me, right now, it does not feel too worrisome, but this might be a lack of experience on the plugin. What we can immediately do: As MSS does not use the counter, we could stop the SQL write, when the sending method is MSS.

I don't have an idea of how many transactional emails are sent by a large site. I will ask in Slack to see if someone has a clue about this.

IMO, it is worth implementing your suggestion to skip incrementing the sending count if MSS is being used just in case. I believe we already do that for emails sent via SendingQueue.

websupporter commented:
> IMO, it is worth implementing your suggestion to skip incrementing the sending count if MSS is being used just in case. I believe we already do that for emails sent via SendingQueue. What do you think?

Addressed in [6aa063d](https://github.com/mailpoet/mailpoet/pull/4001/commits/6aa063d1b8eed236f4b8641dde53fb287203e34c).

I added it straight to the MailerLog::incrementSentCount() method, as:
1. The SendingQueue does not take it into account and increments afaik
2. MailerLog already alters its behavior when the sending method is MSS (regardless of limits, it will return false in isSendingLimitReached, thus has knowledge about this specification)

rodrigoprimo commented:
> The SendingQueue does not take it into account and increments afaik

You are right, I thought it was the SendingQueue but it is actually isSendingLimitReached() where I saw that.

rodrigoprimo reviewed (dismissed):
Looks good to me. Thanks for working on this, @websupporter.

Since this PR touches a core part of MailPoet, I'm going to ask for a second review just to make sure I didn't miss anything. I'm going to assign it to @JanJakes as he already raised an important point on Slack about potential race conditions.

rodrigoprimo commented:
After some discussion on Slack, it was decided that @websupporter will rescope this PR and for now this PR will include just the new data structure to not send over the limits with regards to everything send from queue. We will leave bringing transactional emails into the fold to a separate PR as @JanJakes raised some valid concerns about potential race conditions and it is not obvious to us how we will address it.

JanJakes reviewed (approved):
LGTM, great job! 👍

We'll need to address the race conditions, probably by using full email sending history, as we've discussed.
